Another diamond trial

Jorge-C opened the pull request:
Wanting travis to build it. Ready for reviews.

RNAer committed:
Add Diamond

RNAer committed:
Merge branch 'master' of https://github.com/biocore/conda-recipes

RNAer committed:
Delete unneccessaries

RNAer committed:
Add bc to apt addons

RNAer committed:
Merge branch 'master' of https://github.com/biocore/conda-recipes

RNAer committed:
Use cmake

RNAer committed:
Modify with help from Patrick

Jorge-C committed:
BUG: try to get Diamond working using @patricksnape's work

Jorge-C committed:
BUG: add menpo's channel, patch buggy cmake file, stop calling conda in build script and add boost as dependencies.

Jorge-C committed:
BUG: add patch

Jorge-C committed:
BUG: remove conda build from build.sh

Jorge-C commented:
For the record, I can reproduce the build error from travis in a docker image with 12.04. Not sure how to further debug it, as the error seems to come from some deep function, and there are about three thousand lines of error output.

Jorge-C commented:
It's so much faster to debug stuff in a clean docker image rather than pushing and seeing what travis does...

I've identified the issue: diamond does not compile with gcc/g++ 4.6 (version available in travis/ubuntu 12.04). Updating gcc/g++ to 4.7 solves the problem.

Jorge-C committed:
UPD: update gcc/g++

Jorge-C committed:
Maybe we need to export the updated compilers?

Jorge-C committed:
Full path?

Jorge-C committed:
Pass CC/CXX directly to cmake

Jorge-C committed:
Ops, wrong bash syntax

Jorge-C committed:
The env variables CC/CXX aren't being read?

Jorge-C commented:
I got it working on linux/osx! But some osx expert should check that part, as it's using gcc instead of clang on osx.

If it helps someone else figure out future problems, these are the problems/solutions encountered:
- Having a clean (but local) environment to test these things is super important. I'm not sure if there's currently anything better than docker for that, in particular when dealing with system libraries. By being able to reproduce diamond's build error locally, I figured out that diamond needs gcc >= 4.7.
- Updating compilers in travis. If using the old infrastructure with sudo access, it would have been relatively straightforward

```
before_install:
  - sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
  - sudo apt-get update
  - sudo apt-get install gcc-4.8 g++-4.8
  - sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-4.8 60 --slave /usr/bin/g++ g++ /usr/bin/g++-4.8
  - sudo apt-get update -qq
```

However, the newer container-based travis doesn't have sudo access. It's so much faster that it was worth battling with it a bit more. I managed to install gcc 4.8 easily, but for some reason cmake wasn't finding the new compilers, so after many trials I ended up hard coding them. Suggestions welcome!
- The diamond recipe also needs to patch a bug in diamond (reported [in their repo](https://github.com/bbuchfink/diamond/issues/24)).

ElDeveloper commented:
Just wondering, are you sure you are using gcc and not one of the
symlinks that point to clang?

On (Nov-12-15|20:35), Jorge Cañardo Alastuey wrote:

> I got it working on linux/osx! But some osx expert should check that part, as it's using gcc instead of clang on osx.
> 
> If it helps someone else figure out future problems, these are the problems/solutions encountered:
> - Having a clean (but local) environment to test these things is super important. I'm not sure if there's currently anything better than docker for that, in particular when dealing with system libraries. By being able to reproduce diamond's build error locally, I figured out that diamond needs gcc >= 4.7.
> - Updating compilers in travis. If using the old infrastructure with sudo access, it would have been relatively straightforward
> 
> ```
> before_install:
>   - sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
>   - sudo apt-get update
>   - sudo apt-get install gcc-4.8 g++-4.8
>   - sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-4.8 60 --slave /usr/bin/g++ g++ /usr/bin/g++-4.8
>   - sudo apt-get update -qq
> ```
> 
> However, the newer container-based travis doesn't have sudo access. It's so much faster that it was worth battling with it a bit more. I managed to install gcc 4.8 easily, but for some reason cmake wasn't finding the new compilers, so after many trials I ended up hard coding them. Suggestions welcome!
> - The diamond recipe also needs to patch a bug in diamond (reported [in their repo](https://github.com/bbuchfink/diamond/issues/24)).
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/conda-recipes/pull/41#issuecomment-156319433

Jorge-C commented:
I don't know, that's why I'd like someone who uses osx to double check that part. The following output is coming from cmake in the OSX travis build:

```
-- The C compiler identification is GNU 4.8.3
-- The CXX compiler identification is GNU 4.8.3
```

ElDeveloper commented:
Got it, thanks for the info. Thinking a bit more about it, it should
just work ... though I could very well be wrong.

On (Nov-13-15|10:47), Jorge Cañardo Alastuey wrote:

> I don't know, that's why I'd like someone who uses osx to double check that part. The following output is coming from cmake in the OSX travis build:
> 
> ```
> -- The C compiler identification is GNU 4.8.3
> -- The CXX compiler identification is GNU 4.8.3
> ```
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/conda-recipes/pull/41#issuecomment-156518915

RNAer commented on .travis.yml:
```diff
@@ -8,7 +15,9 @@ env:
-    secure: gKgj4jR2N8WbVBnJCI52zMR21/nd2hWUb05DatiWNJ8pZ+CC6CPD7N/2t8k+y5V25NufI5ZpEBEnGbmV1AxS6ISjPHRejoQgDYTw/PjY4NEIssFVbULIB84FD9SY7/oqkr5L+jrEyDeV52d2LfSAHjwWlL4XYZ7OFdH7HW/5yUj1z4jasfFhv41x8tVknXWAeS1AEdAfwnb4lEeKvlP6jxdfqkQv9PeobM+THV7gGtUo9rA+plZuhQDzXhywRKajEbYbA7/eHzQa5kMYtf8EQl/orDOPCSMEjh9ifCsf0bcFuQkqtOKhgPGVihNdw9PErwhxAckiH2lyELilqbbk10VI/kTKb8y1yP+RmaRxHD1TzdUL5qAxlSQx8IVybnmKCgB ...
+    - secure: gKgj4jR2N8WbVBnJCI52zMR21/nd2hWUb05DatiWNJ8pZ+CC6CPD7N/2t8k+y5V25NufI5ZpEBEnGbmV1AxS6ISjPHRejoQgDYTw/PjY4NEIssFVbULIB84FD9SY7/oqkr5L+jrEyDeV52d2LfSAHjwWlL4XYZ7OFdH7HW/5yUj1z4jasfFhv41x8tVknXWAeS1AEdAfwnb4lEeKvlP6jxdfqkQv9PeobM+THV7gGtUo9rA+plZuhQDzXhywRKajEbYbA7/eHzQa5kMYtf8EQl/orDOPCSMEjh9ifCsf0bcFuQkqtOKhgPGVihNdw9PErwhxAckiH2lyELilqbbk10VI/kTKb8y1yP+RmaRxHD1TzdUL5qAxlSQx8IVybnmKC ...
+    - CC=gcc-4.8
+    - CXX=g++-4.8
```
`CC` and `CXX` are still needed? specifying the version globally may interfere the building of other packages?

Jorge-C replied:
:+1: I'll drop them

RNAer commented on ci/travis/install.sh:
```diff
@@ -21,3 +21,7 @@ conda update -q conda
+
+# Add channel to get boost libraries. Probably want to "self-host",
+# maybe embedding their recipes as a git module?
+conda config --add channels menpo
```
This is not put in `build.sh` because you think `boost` may be required for other packages too?

Jorge-C replied:
Maybe, but it has to be here because there can't be calls to conda inside the build script.

RNAer replied:
why is that? I think I ran it inside `build.sh`: https://travis-ci.org/biocore/conda-recipes/jobs/90782083#L548

Jorge-C replied:
That build failed. Once I got building working, installing failed because of the conda call. Luckily, it's easy to understand b/c the build script is not called when installing, which means that boost isn't there when trying to run diamond and things fall apart.

tl;dr Nothing can be installed inside build scripts by calling conda, they must be set as dependencies in the meta file.

Jorge-C replied:
Note that this is still suboptimal: to install diamond our users would have to add two channels to their configs, which I think hampers usability. To solve that, we need to either get it working with the default boost, or host the menpo recipes and keep them in the biocore channel (as suggested by Patrick).

RNAer replied:
I see.

Totally agree. If not too much trouble, fixing the default boost is better; otherwise, host menpo recipe as a submodule is also fine.

RNAer commented on recipes/Diamond/cmake_patch:
```diff
@@ -0,0 +1,19 @@
+-target_link_libraries(diamond blast_core ${Boost_LIBRARIES})
++target_link_libraries(diamond blast_core ${Boost_LIBRARIES} ${CMAKE_THREAD_LIBS_INIT})
+ 
+ install(TARGETS diamond DESTINATION bin)
```
:+1:

Jorge-C committed:
Remove CC/CXX as they're not needed

Jorge-C commented:
@RNAer I addressed your comments, travis is green.

RNAer commented:
I thought you'd like to take a stab on the default boost or put menpo boost
as submodule? if not, I can merge it.

On Tue, Nov 17, 2015 at 12:21 PM, Jorge Cañardo Alastuey <
notifications@github.com> wrote:

> @RNAer https://github.com/RNAer I addressed your comments, travis is
> green.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/biocore/conda-recipes/pull/41#issuecomment-157495633.

Jorge-C commented:
Yeah, but I won't have time to work on it for a bit so maybe merge and look
into it at a later time. I don't have any strong preference.
On Nov 17, 2015 1:58 PM, "Zech Xu" notifications@github.com wrote:

> I thought you'd like to take a stab on the default boost or put menpo boost
> as submodule? if not, I can merge it.
> 
> On Tue, Nov 17, 2015 at 12:21 PM, Jorge Cañardo Alastuey <
> notifications@github.com> wrote:
> 
> > @RNAer https://github.com/RNAer I addressed your comments, travis is
> > green.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/biocore/conda-recipes/pull/41#issuecomment-157495633
> > .
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/biocore/conda-recipes/pull/41#issuecomment-157505263.

RNAer commented:
There are still problems even it seems built successfully: https://travis-ci.org/biocore/micronota/jobs/91713275

Also error on local Mac:

```
$ conda install -c biocore diamond
$ diamond
dyld: Symbol not found: __ZNK5boost15program_options22error_with_option_name23substitute_placeholdersERKSs
  Referenced from: /Users/zech/anaconda/envs/py3/bin/diamond
  Expected in: /Users/zech/anaconda/envs/py3/bin/../lib//./libboost_program_options.dylib
 in /Users/zech/anaconda/envs/py3/bin/diamond
Trace/BPT trap: 5
```
